### PR TITLE
Allow Navigator to use authenticated Ledger API

### LINKED ISSUE #3156: Navigator: read and use access token
Navigator should read the access token from `$DAML_HOME/secret` (or a configurable alternative location) and use it to work against an authenticated sandbox.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Fixes #3156 

Also removes legacy usage of password workflow and exposes new convenience methods to create a `LedgerClient` without having to re-write the logic to create a channel.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala [hunk 1]
```diff
@@ -77,38 +80,73 @@ object LedgerClient {
...
+  def constructChannel(
+      hostIp: String,
+      port: Int,
+      configuration: LedgerClientConfiguration): Try[Channel] = {
+    val builder: NettyChannelBuilder = NettyChannelBuilder.forAddress(hostIp, port)
     configuration.sslContext
       .fold {
-        builder.usePlaintext()
+        if (configuration.accessToken.isDefined) {
+          Failure(new IllegalArgumentException("Cannot send a token over an insecure channel"))
```
Comment (rautenrieth-da) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: Should we provide a way to use tokens on insecure channels for development purposes?

I like the strict check here, provided that we have an easy way of working with a local sandbox in SSL mode (easy meaning the user does not have to manually run SSL certificate generating commands).

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala [hunk 2]
```diff
@@ -77,38 +80,73 @@ object LedgerClient {
...
+    val creds = configuration.accessToken.map(callCredentials).orNull
     apply(
-      LedgerIdentityServiceGrpc.stub(channel),
-      TransactionServiceGrpc.stub(channel),
-      ActiveContractsServiceGrpc.stub(channel),
-      CommandSubmissionServiceGrpc.stub(channel),
-      CommandCompletionServiceGrpc.stub(channel),
-      CommandServiceGrpc.stub(channel),
-      PackageServiceGrpc.stub(channel),
+      LedgerIdentityServiceGrpc.stub(channel).withCallCredentials(creds),
```
Comment (rautenrieth-da) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: I assume `.withCallCredentials(null)` is a no-op?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala [hunk 3]
```diff
@@ -90,7 +96,12 @@ class SandboxServer(actorSystemName: String, config: => SandboxConfig) extends A
   // TODO: Pass this info in command-line (See issue #2025)
   val participantId: ParticipantId = Ref.LedgerString.assertFromString("sandbox-participant")
 
-  private val authService: AuthService = AuthServiceWildcard
+  // FIXME If this code is still here when reviewing, please ask me to remove it
+  private val authService: AuthService = AuthServiceStatic {
+    case "root" => Claims.wildcard
+    case "admin" => Claims(Seq(ClaimAdmin))
```
Comment (rautenrieth-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: Add `ClaimPublic`, otherwise you won't be able to use "public" APIs. Same below.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala (hunk 1 above)
Comment (rautenrieth-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: I see you added some certificates in `navigator/backend/scenarios/rental/tls/` :+1:

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala [hunk 4]
```diff
@@ -56,6 +56,7 @@ abstract class UIBackend extends LazyLogging with ApplicationInfoJsonSupport {
     * (supply a default)
     */
   def defaultConfigFile: Path = Paths.get("ui-backend.conf")
+  def defaultAccessTokenFile: Path = Paths.get(sys.props("user.home"), ".daml", "secret")
```
Comment (rautenrieth-da) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: The location of this file could be read using the `ProjectConfig` library.

See https://github.com/digital-asset/daml/blob/master/daml-assistant/scala-daml-project-config/src/main/scala/com/digitalasset/assistant/config/ProjectConfig.scala for the scala library (which currently only supports the project config and not the global daml config).

See https://github.com/digital-asset/daml/blob/master/daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs#L43 and https://github.com/digital-asset/daml/blob/master/daml-assistant/daml-project-config/DA/Daml/Project/Config.hs#L38 for the original Haskell library.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: It explicitly sets the `CallCredentials` to be `null`, so that they are not used.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: Ah, I thought public was equal to non-empty. Shouldn't we make it that way?

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: As @gerolf-da mentioned offline, we should probably not do it here. The thing I was mostly interested in was providing a clear error message to the user. I'll see what happens if I try to do it without any such protection. IIRC gRPC itself should complain, I'll try to see if it's possible to get back some meaningful error to return to the user.

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: `ProjectConfig` is about the `daml.yaml` file, where we should not store the token (since that is supposed to be checked in with the rest of your code).

After a short conversation with @associahedron and @hurryabit we agreed on using `$DAML_HOME/secret` as a well-known location (`$DAML_HOME` is `~/.daml` on Mac/Linux), while leaving the possibility to pass a different location on the command line.

On top of that, I'll probably make an effort to try to read from the expected well-known location if `$DAML_HOME` is not found, unless you have something against it.

So basically something along these lines:
```scala
// Uses $DAML_HOME from the SDK Assistant or falls back to its default location:
// - on Windows: %APPDATA%/daml
// - on Mac and Linux: ~/.daml
// The file called `secret` will be looked in the resolved location
private[this] val defaultAccessTokenFile: Path =
  sys.env
    .get("DAML_HOME")
    .fold(if (sys.props("os.name").toLowerCase.startsWith("win")) {
      Paths.get(sys.env("APPDATA"), "daml")
    } else {
      Paths.get(sys.props("user.home"), ".daml")
    })(Paths.get(_))
    .resolve("secret")
```

File: navigator/backend/scenarios/README.md [hunk 5]
```diff
@@ -24,4 +24,20 @@ As an example, assuming your current working directory is the project root, you
 
 1. open a browser
 1. go to `localhost:4000`
-1. play around
\ No newline at end of file
+1. play around
+
+#### Usage with TLS
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply navigator/backend/scenarios/README.md]: This is only for the test setup for contributors to this repo, I'll add actual docs later on. ;)

**Comment by stefanobaghino-da** [comment]
@bame-da @gerolf-da The docs are not meant as extensive documentation for authentication (of course) but just on how to use Navigator against a Ledger API that requires authentication. Documentation for authentication as its own topic will come in a separate PR.

File: docs/source/tools/navigator/index.rst [hunk 6]
```diff
@@ -247,6 +247,24 @@ To advance time:
   :width: 25%
   :align: center
 
+Authenticating Navigator
+************************
+
+If you are running Navigator against a Ledger API server that requires authentication, you must provide the access token when you start the Navigator server.
+
+The access token retrieval depends on the specific DAML setup you are working with: please refer to the ledger operator to learn how.
```
Comment (rautenrieth-da) [inline docs/source/tools/navigator/index.rst]: Maybe also mention the exact content of the file, like whether it will have `Bearer ` prepended?

File: navigator/backend/src/main/scala/com/digitalasset/navigator/config/Arguments.scala [hunk 7]
```diff
@@ -69,29 +72,29 @@ object Arguments {
...
 
       opt[String]("assets")
         .text("folder where frontend assets are available")
         .action((assets, arguments) => arguments.copy(assets = Some(assets)))
 
-      opt[Unit]('w', "require-password")
-        .text("if true then the password is required to login")
-        .action((_, arguments) => arguments.copy(requirePassword = true))
+      opt[String]('t', "access-token-file")
+        .text(s"set the access token file, default: ${defaultAccessTokenFile}")
```
Comment (rautenrieth-da) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/config/Arguments.scala]: Do we need something like `See the the docs for how to generate the token file`, or is that obvious?

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala [hunk 8]
```diff
@@ -57,12 +61,27 @@ abstract class UIBackend extends LazyLogging with ApplicationInfoJsonSupport {
     */
   def defaultConfigFile: Path = Paths.get("ui-backend.conf")
 
+  // Uses $DAML_HOME from the SDK Assistant or falls back to its default location:
+  // - on Windows: %APPDATA%/daml
+  // - on Mac and Linux: ~/.daml
+  // The file called `secret` will be looked in the resolved location
+  private[this] val defaultAccessTokenFile: Path =
+    sys.env
+      .get("DAML_HOME")
```
Comment (gerolf-da) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: I'm wondering whether we should put this logic into the general daml.yaml (or rather daml-project-config) library?

File: navigator/backend/src/test/scala/com/digitalasset/navigator/backend/ServerTest.scala [hunk 9]
```diff
@@ -99,45 +109,18 @@ class ServerTest
   }
 
   it should "forbid to SignIn with a non existing user" in withCleanSessions {
```
Comment (rautenrieth-da) [inline navigator/backend/src/test/scala/com/digitalasset/navigator/backend/ServerTest.scala]: At some point, we should probably also have a `it should "forbid to SignIn while still connecting to a ledger"` and `it should "forbid to SignIn when permanently failed to connect"`

File: navigator/frontend/src/ui-core/src/session/UI.tsx [hunk 10]
```diff
@@ -85,15 +102,32 @@ export default class Component<A extends Action>
     let errorEl = null;
     if (failure === 'invalid-credentials') {
       errorEl = (
-        <p>Sorry, those credentials are invalid</p>
+        <ErrorMessage>
+          <div>You don't have the necessary authorization to access</div>
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/session/UI.tsx]:
```suggestion
          <div>You don't have the necessary authorization to access the ledger</div>
```

Sounds like there is an object missing at the end?

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala [hunk 11]
```diff
@@ -119,44 +128,41 @@ abstract class UIBackend extends LazyLogging with ApplicationInfoJsonSupport {
...
-                                  if userConfig.password == None || userConfig.password.contains(
-                                    password) =>
+                      config.users.get(request.userId) match {
+                        case None =>
+                          logger.error(
+                            s"Attempt to signin with non-existent user ${request.userId}")
+                          complete(signIn(Some(InvalidCredentials)))
+                        case Some(userConfig) =>
+                          import system.dispatcher
+                          Await.result(
```
Comment (gerolf-da) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: Instead of `Await.result` you can use the `onSuccess` directive.

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 8 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: `daml-project-config` is about the project-specific configuration, if we make a library of it it should be something different. Since I'll for sure need this piece of code to make Extractor work with authentication I'd suggest we defer it to that contribution. Sounds good?

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 11 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: I didn't know about `akka-http`'s `onSuccess`, thanks for the tip!

File: navigator/frontend/src/ui-core/src/session/UI.tsx (hunk 10 above)
Comment (stefanobaghino-da) [reply navigator/frontend/src/ui-core/src/session/UI.tsx]: This line was on purpose but ok.

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: I'll use `onSuccess` as suggested by @gerolf-da.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala [hunk 12]
```diff
@@ -77,38 +79,61 @@ object LedgerClient {
...
-      .fold {
-        builder.usePlaintext()
-      } { sslContext =>
-        builder.sslContext(sslContext).negotiationType(NegotiationType.TLS)
+      esf: ExecutionSequencerFactory): Future[LedgerClient] =
+    forChannel(configuration, constructChannel(hostIp, port, configuration))
+
+  private[this] val auth = Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER)
+
+  def callCredentials(token: String): CallCredentials =
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: I'll make the `Bearer` part optional.

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 11 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: Addressed by https://github.com/digital-asset/daml/pull/3138/commits/609adf670962eca159a75ff0e74773dc7930aca2

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/LedgerClient.scala]: Addressed by https://github.com/digital-asset/daml/pull/3138/commits/827472c54d38756f8d14b02b0b7640fd5ad6b3cd

File: docs/source/tools/navigator/index.rst (hunk 6 above)
Comment (stefanobaghino-da) [reply docs/source/tools/navigator/index.rst]: That is specific to the setup, pointing the reader to the ledger operator is the only thing we can do reliably.

File: navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala (hunk 8 above)
Comment (gerolf-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/UIBackend.scala]: Sounds good. 👍

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala [hunk 13]
```diff
@@ -9,4 +9,5 @@ final case class LedgerClientConfiguration(
     applicationId: String,
     ledgerIdRequirement: LedgerIdRequirement,
     commandClient: CommandClientConfiguration,
-    sslContext: Option[SslContext])
+    sslContext: Option[SslContext],
+    accessToken: Option[String] = None)
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala]: Addressed by https://github.com/digital-asset/daml/pull/3138/commits/036f8bea1a07eedc4fe4c365c27d1c062e8d4910

File: navigator/backend/src/main/scala/com/digitalasset/navigator/config/Arguments.scala (hunk 7 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/config/Arguments.scala]: Addressed by https://github.com/digital-asset/daml/pull/3138/commits/ef7a382a50e0bd6dd3e34d72318e6979fb9df764

File: navigator/backend/src/test/scala/com/digitalasset/navigator/backend/ServerTest.scala (hunk 9 above)
Comment (stefanobaghino-da) [reply navigator/backend/src/test/scala/com/digitalasset/navigator/backend/ServerTest.scala]: Some point is https://github.com/digital-asset/daml/pull/3138/commits/588e11b6524b1b6e66d5a6bd82dd5e1729484974 :slightly_smiling_face: 

Very good point, thanks for bringing this up.
